tracer/src/Datadog.Trace/Tagging/TraceTagCollection.cs

```diff
@@ -0,0 +1,189 @@
+// <copyright file="TraceTagCollection.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using Datadog.Trace.Util;
+
+namespace Datadog.Trace.Tagging
+{
+    internal class TraceTagCollection
+    {
+        // key1=value1,key2=value2
+        private const char TagPairSeparator = ',';
+        private const char KeyValueSeparator = '=';
+
+        // tags with this prefix are propagated horizontally
+        // (i.e. from upstream services and to downstream services)
+        // using the `x-datadog-tags` header
+        private const string PropagatedTagPrefix = "_dd.p.";
+
+        // for now we only expect one trace-level tag:
+        // "_dd.p.upstream_services"
+        private const int DefaultCapacity = 1;
+
+        // "_dd.p.a=b"
+        public const int MinimumPropagationHeaderLength = 9;
+        public const int MaximumPropagationHeaderLength = 512;
+
+        private static readonly char[] TagPairSeparators = { TagPairSeparator };
+
+        private List<KeyValuePair<string, string>> _tags;
+        private string? _cachedPropagationHeader;
+
+        public TraceTagCollection()
+        {
+            _tags = new List<KeyValuePair<string, string>>(DefaultCapacity);
+        }
+
+        public TraceTagCollection(List<KeyValuePair<string, string>> tags)
+        {
+            _tags = tags;
+        }
+
+        public static TraceTagCollection ParseFromPropagationHeader(string? propagationHeader)
+        {
+            if (string.IsNullOrEmpty(propagationHeader))
+            {
+                return new TraceTagCollection();
+            }
+
+            var tags = propagationHeader!.Split(TagPairSeparators, StringSplitOptions.RemoveEmptyEntries);
+            var tagList = new List<KeyValuePair<string, string>>(tags.Length);
+
+            foreach (var tag in tags)
+            {
+                var separatorIndex = tag.IndexOf(KeyValueSeparator);
```

Comment:
Out of curiosity, we filter when we call `FormatPropagationHeader()` the tags that do not start by `_dd.p.`, why don't we do the same here? 
I get that it shouldn't happen anyway. But this also made me realize that we don't have the same behavior if we set a tag and if we don't. If we don't, we reuse the cached string so we forward anything. If we do set a tag, then we recompute the headers and potentially filter out the non `_dd.p` one. Should we not filter out when we call `FormatPropagationHeader()` ?

Reply:
Yes, thank you! I added more checks to identify propagated tags (with `_dd.p.*` prefix) vs non-propagated tags. Only trace tags with `_dd.p.*` are propagated horizontally via the `x-datadog-tags` header, but other trace tags are also supported too (they are still sent to the Agent and may be propagated via a different header, like `x-datadog-sampling-priority`).

I should probably clarify this distinction in the comments...

Reply:
I added more comments to clarify that we ignore tags without `_dd.p.*` when parsing and serializing the `x-datadog-tags` header.